### PR TITLE
Created LinkedList and Node classes

### PR DESCRIPTION
**Pull request by TreverArnold**
Also defined basic operation functions, basic tests have been completed through sub-par methods, I will be adding a test file and using pytest upon next commit.

### CODE REVIEW

**Comment by W-Ely** [comment]
Your readme should look something like:
```markdown
# Data Structures
Holds sample code for a number of classic data structures implemented in Python.

## LinkedList:
Module: [path/to/code.py](path/to/code.py)
Tests:[path/to/test.py](path/to/test.py)
    * push(val): Inserts the value at the head of the list
    * pop(): Pops the first value off the head of the list and returns it. Raises an exception if there are no values to return.
    * size(): Returns the length of the list.
    * search(val): Returns the node containing the val in the list if it is present. If it is not present it will return None.
    * remove(node): Removes the given node from the list. If the node is not in the list is will raise an exception.
    * display(): Returns a unicode string representing the list as if it were a Pthon tuple literal. Ex: "(5, 'Bravo', 9)"
    * len(the_list): Returns the size of the list
    * print(the_list): Returns what the display method returns.
```

**Comment by W-Ely** [comment]
Will further review once the tests are present.

File: linked_list.py [hunk 1]
```diff
@@ -13,6 +14,8 @@ def __init__(self, iterable=None):
                 self.push(item)
 
     def push(self, value):
+        if value is None or (isinstance(value, str) and not value.strip()):
+            raise ValueError("Push needs a value")
```
Comment (W-Ely) [inline linked_list.py]: An empty string should be an acceptable value.

File: linked_list_test.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    assert my_list._size == 3
+    assert my_list.head.value == 'hello'
+    assert my_list.head.next.value == 2
+    assert my_list.head.next.next.value == 1
+    assert my_list.head.next.next.next == None
+
+def test_linked_list_push_empty():
+    with pytest.raises(ValueError):
+        my_list = LinkedList([1, 2])
+        my_list.push('')
```
Comment (W-Ely) [inline linked_list_test.py]: Pushing an empty string is perfectly fine. Nice use of `with pytest.raises(ValueError):` though!

File: linked_list_test.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+def test_linked_list_display_empty():
+    my_list = LinkedList()
+    assert my_list.display() == "()"
+
+def test_linked_list_display():
+    my_list = LinkedList([1, 2, 3, 4, 5])
+    assert my_list.display() == "(5, 4, 3, 2, 1)"
+    linked_list = LinkedList([1, 2, 3, 4, 5])
+    print(linked_list)
+    captured_output = capture_output_for_print_test(print)(linked_list)
```
Comment (W-Ely) [inline linked_list_test.py]: Line 42 isn't doing anything here. Also this test shouldn't be testing `print` it's the test for Display and display does not print anything.

File: linked_list_test.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            output_buffer.close()
+
+        return captured_output
+
+    return wrapper
+
+def test_linked_list_print():
+    linked_list = LinkedList([1, 2, 3, 4, 5])
+    print(linked_list)
+    captured_output = capture_output_for_print_test(print)(linked_list)
```
Comment (W-Ely) [inline linked_list_test.py]: This is the only place you need to use the function `capture_output_for_print_test` so I'd rather you remove it and put the logic here that actually tests print.  Line 135 isn't being used, but should be.  Hint some of `capture_output_for_print_test` goes before line 135 and some of it after.

File: linked_list_test.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        try:
+            func(*args, **kwargs)
+        finally:
+            sys.stdout = sys.__stdout__
+            captured_output = output_buffer.getvalue()
+            output_buffer.close()
+
+        return captured_output
+
+    return wrapper
```
Comment (W-Ely) [inline linked_list_test.py]: This is a fairly advanced usage, which is fine, but not needed since you only have one test here that should care what is printed.  Anytime you `wrap` a function you should use functools.wraps https://towardsdatascience.com/why-you-should-wrap-decorators-in-python-5ac3676835f9

Though your uses of it above and below are unorthodox. Typically when one creates a `decorator` like you have here. The return would be what the func on line 123 returns.  It's clear why you aren't, you want the captured output, but it also means you can't use this decorator with the Python `@` syntax that one would expect. 

Also within a test file only tests should have the word `test` in them.  It will confuse Pytest and might lead to unexpected outcomes.

File: linked_list_test.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    expected_output = "(5, 4, 3, 2, 1)"
+    assert captured_output.strip() == expected_output
+
+def test_linked_list_pop():
+     my_list = LinkedList([3, 4, 5])
+     pop1 = my_list.pop()
+     pop2 = my_list.pop()
+     pop3 = my_list.pop()
+     pops = repr(pop1) + ", " + repr(pop2) + ", " + repr(pop3)
+     assert pops == '5, 4, 3'
```
Comment (W-Ely) [inline linked_list_test.py]: Instead of doing the repr and testing what the output string looks like check the actual values.
```suggestion
     assert (pop1, pop2, pop3) == (5, 4, 3)
```

File: linked_list_test.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+def test_linked_list_pop_empty():
+     with pytest.raises(ValueError):
+        my_list = LinkedList([])
+        my_list.pop()
```
Comment (W-Ely) [inline linked_list_test.py]: You shouldn't need to supply the empty list when creating an empty linked list.

How about we make sure the Error is the correct value error. Something like:
```suggestion
     with pytest.raises(ValueError) as error:
        my_list = LinkedList()
        my_list.pop()
    assert error.value.args[0] == "Pop does not work on empty array or list"
```

File: linked_list.py [hunk 8]
```diff
@@ -25,28 +28,28 @@ def display(self):
             elements.append(repr(current.value))
             current = current.next
         return "(" + ", ".join(elements) + ")"
-    
+
     def pop(self):
         if self.head is None:
-            raise Exception("Pop does not work on empty array or list")
-        
+            raise ValueError("Pop does not work on empty array or list")
```
Comment (W-Ely) [inline linked_list.py]:
```suggestion
            raise ValueError("Pop does not work on an empty list")
```

File: linked_list_test.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
+from linked_list import LinkedList
+import io
+import sys
+import pytest
```
Comment (W-Ely) [inline linked_list_test.py]: Follow pep 0008 conventions for ordering your imports

File: linked_list_test.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+def test_linked_list_remove():
+    my_list = LinkedList([1, 2, 3, 4, 5])
+    my_list.remove(my_list.head.next.next)
+    assert my_list.head.next.next.value == 2
+
+def test_linked_list_remove_non_existent():
+    with pytest.raises(ValueError):
+        my_list = LinkedList([1, 2, 3, 4])
+        my_list.remove(5)
```
Comment (W-Ely) [inline linked_list_test.py]: Check that the correct error message appears here too.

File: linked_list_test.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    my_list = LinkedList([1, 2, 3, 4, 5])
+    assert len(my_list) == 5    
+    my_list = LinkedList([1, 2, 3, 4])
+    assert len(my_list) == 4    
+    my_list = LinkedList([1, 2, 3])
+    assert len(my_list) == 3    
+    my_list = LinkedList([1, 2])
+    assert len(my_list) == 2    
+    my_list = LinkedList([1])
+    assert len(my_list) == 1
```
Comment (W-Ely) [inline linked_list_test.py]: What logic has the highest chance of causing error s with your length?  Probably not just instantiating the list which only  calls the `__init__` function, but rather any place that updates your `_size` property.
This would be better:
```suggestion
def test_linked_list_len():
    my_list = LinkedList([1, 2, 3])
    assert len(my_list) == 3    
    my_list.pop()
    assert len(my_list) == 2
    my_list.pop()
    assert len(my_list) == 1    
    my_list.pop()
    assert len(my_list) == 0 
    my_list.push(1)
    assert len(my_list) == 1
    my_list.push(2)
    assert len(my_list) == 2
    my_list.push(3)
    assert len(my_list) == 3
    my_list.remove(my_list.head.next)
    assert len(my_list) == 2
    my_list.remove(my_list.head.next)
    assert len(my_list) == 1
    my_list.remove(my_list.head)
    assert len(my_list) == 0
```

File: linked_list_test.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+def test_linked_list_str():
+    my_list = LinkedList([1, 2, 3])
+    assert str(my_list) == '(3, 2, 1)'
+    assert str(my_list.head.value) == '3'
+    assert str(my_list.head.next.value) == '2'
+    assert str(my_list.head.next.next.value) == '1'
+    assert str(my_list.head.next.next.next) == 'None'
```
Comment (W-Ely) [inline linked_list_test.py]: Line 99 is the only meaningful assertion here.  It is testing the `__str__ ` of the linked list.  What are lines 100 - 103 testing?

File: linked_list_test.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+def test_linked_list_size():
+    my_list = LinkedList([1, 2, 3, 4, 5])
```
Comment (W-Ely) [inline linked_list_test.py]: The size test should follow the same pattern as the length test, covering all cases that might effect the size of the list.  From empty to not empty and from not empty to empty are always especially import since these are usually the edge cases.  Testing from 4 to 5 isn't really any different than testing from 3 to 4.

**Review by W-Ely — CHANGES_REQUESTED** [review]
Great Work! Nice finds and lots of opportunities for improvement.

File: linked_list_test.py (hunk 3 above)
Comment (TreverArnold) [reply linked_list_test.py]: When I was re-arranging the order of the tests, I made a few mistakes, this appears to be the remnant of one.

File: linked_list_test.py (hunk 12 above)
Comment (TreverArnold) [reply linked_list_test.py]: That is a good question.

File: linked_list_test.py (hunk 4 above)
Comment (TreverArnold) [reply linked_list_test.py]: When writing this section I was beyond confused on what I was even doing, this hint along with a bit of research has helped a lot. Thank you.

File: linked_list_test.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
+import io
+import sys
+import pytest
+from linked_list import LinkedList
```
Comment (W-Ely) [inline linked_list_test.py]: [pep 0008 ](https://peps.python.org/pep-0008/#imports)
```suggestion
import io
import sys

import pytest

from linked_list import LinkedList
```

File: linked_list_test.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+     my_list = LinkedList([3, 4, 5])
+     pop = my_list.pop()
+     assert pop == 5
+     pop = my_list.pop()
+     assert pop == 4
+     pop = my_list.pop()
+     assert pop == 3
+
+def test_linked_list_pop_empty():
+     with pytest.raises(ValueError, match="Pop does not work on an empty list"):
```
Comment (W-Ely) [inline linked_list_test.py:53]: This is better than my suggestion.

**Review by W-Ely — COMMENTED** [review]
So close!

**Review by W-Ely — APPROVED** [review]
